File: src/go/coordinate.py
#!/usr/bin/env python3

from .constants import MAX_SPAN
from typing import List


class Coordinate:
    def __init__(self, row: int, column: int):
        self.row = row
        self.column = column
        self.neighbors: List[Coordinate] = []
        self.corners: List[Coordinate] = []
        self._hash = hash((self.row, self.column))

    def __deepcopy__(self, memodict = {}):
        return self

    def __str__(self):
        return f'{self.column:02}-{self.row:02}'

    def __eq__(self, other):
        return other is self or isinstance(other, Coordinate) and self.row == other.row and self.column == other.column

    def __ne__(self, other):
        return not self.__eq__(other)

    def __hash__(self):
        return self._hash

class Coordinates:
    def __init__(self, span: int):
        assert 1 <= span <= MAX_SPAN
        self.span = span
        self._coordinates = [Coordinate(row, column) for row in range(1, span + 1) for column in range(1, span + 1)]
        for coordinate in self._coordinates:
            for pair in [
                (coordinate.row - 1, coordinate.column),
                (coordinate.row, coordinate.column - 1),
                (coordinate.row, coordinate.column + 1),
                (coordinate.row + 1, coordinate.column)
            ]:
                row, column = pair
                if self._validate_coordinate(row, column):
                    index = self._calculate_index(row, column)
                    other = self._coordinates[index]
                    coordinate.neighbors.append(self._coordinates[index])
            for pair in [
                (coordinate.row - 1, coordinate.column - 1),
                (coordinate.row - 1, coordinate.column + 1),
                (coordinate.row + 1, coordinate.column - 1),
                (coordinate.row + 1, coordinate.column + 1)
            ]:
                row, column = pair
                if self._validate_coordinate(row, column):
                    index = self._calculate_index(row, column)
                    other = self._coordinates[index]
                    coordinate.corners.append(other)

    def _validate_coordinate(self, row: int, column: int):
        return self._validate_component(row) and self._validate_component(column)

    def _validate_component(self, component):
        return 1 <= component <= self.span

    def _calculate_index(self, row: int, column: int):
        return (row - 1) * self.span + column - 1

    def get(self, row: int, column: int):
        self._validate_coordinate(row, column)
        index = self._calculate_index(row, column)
        return self._coordinates[index]

    def __iter__(self):
        return iter(self._coordinates)

    def __contains__(self, coordinate: Coordinate):
        return coordinate and self._validate_coordinate(coordinate.row, coordinate.column)


_coordinates = {}


def get_coordinates(span: int):
    assert 1 <= span <= MAX_SPAN
    if span not in _coordinates:
        _coordinates[span] = Coordinates(span)
    return _coordinates[span]


PASS = Coordinate(-1, -1)

File: src/go/coordinate.c
#include <stdbool.h>
#include <string.h>
#include "coordinate.h"

static bool initialized = false;
static Coordinates cache[MAX_SPAN] = {};

Coordinates *get_coordinates(size_t span) {
    // Zero out all the arrays in the cache if this is the first call.
    if (!initialized) {
        memset(cache, 0, sizeof cache);
        initialized = true;
    }

    // Return NULL if the requested size is invalid.
    if (span < 1 || span > MAX_SPAN) {
        return NULL;
    }

    // Populate the requested size if it has not been populated before.
    Coordinates *coordinates = &(cache[span - 1]);
    if (coordinates->span != span) {
        // Set the core fields.
        const size_t area = span * span;
        coordinates->span = span;
        coordinates->count = area;

        // Set all the basic data for each Coordinate.
        for (size_t i = 0; i < area; ++i) {
            Coordinate *coordinate = &(coordinates->coordinates[i]);
            coordinate->span = span;
            coordinate->index = i;
            coordinate->row = i / span + 1;
            coordinate->column = i % span + 1;
            coordinate->neighbor_count = 0;
            coordinate->corner_count = 0;
            memset(&(coordinate->neighbors), 0, 4 * sizeof(Coordinate *));
            memset(&(coordinate->corners), 0, 4 * sizeof(Coordinate *));
        }

        // Now that all the Coordinates are prepared, fill in their neighbor and corner arrays.
        for (size_t row = 1, i = 0; row <= span; ++row) {
            for (size_t column = 1; column <= span; ++i, ++column) {
                Coordinate *coordinate = &(coordinates->coordinates[i]);

                bool has_bottom = row + 1 <= span;
                if (column + 1 <= span) {
                    Coordinate *neighbor = &(coordinates->coordinates[i + 1]);
                    coordinate->neighbors[coordinate->neighbor_count++] = neighbor;
                    neighbor->neighbors[neighbor->neighbor_count++] = coordinate;

                    if (row > 1) {
                        Coordinate *corner = &(coordinates->coordinates[i - span + 1]);
                        coordinate->corners[coordinate->corner_count++] = corner;
                        corner->corners[corner->corner_count++] = coordinate;
                    }

                    if (has_bottom) {
                        Coordinate *corner = &(coordinates->coordinates[i + span + 1]);
                        coordinate->corners[coordinate->corner_count++] = corner;
                        corner->corners[corner->corner_count++] = coordinate;
                    }
                }
                if (has_bottom) {
                    Coordinate *neighbor = &(coordinates->coordinates[i + span]);
                    coordinate->neighbors[coordinate->neighbor_count++] = neighbor;
                    neighbor->neighbors[neighbor->neighbor_count++] = coordinate;
                }
            }
        }
    }

    // Return the requested Coordinates.
    return coordinates;
}
